### ISSUE TITLE
Playlist are doubled

### ISSUE THREAD

**Issue by Ilya87** [MBach/Miam-Player#33]
I tried different combinations of save playlists options and found out that I can't find option that restore only playlists I had opened before exit. Current behaviours is - save playlist, restore them after you exit/start program, then Miam Player repeats this procedure (remember before the exit we already have 2 saved playlists) for new exit/start cycle - and we have 4 playlist, etc...

**Comment by MBach** [comment]
Hash computation of playlist content needs some review. Also, the icon left to playlist's name should reflect when one has changed a playlist.

**Comment by MBach** [comment]
Ok so https://github.com/MBach/Miam-Player/commit/fd4ce7bc34cddd79fea5576e2eb829a43a680c3c (and previous) should greatly improve how playlists are managed!

**Comment by Ilya87** [comment]
Now options save/load playlist doesn't work at all. After exit and then start - blank playlist appears.

**Comment by MBach** [comment]
Did you try with newly created playlists ?

**Comment by Ilya87** [comment]
Yes, I (as usual) have deleted all Miam-Player's config folders, created new playlists, quit, started again - only single blank playlist is available, no saved plalists in playlist manager.

**Comment by MBach** [comment]
Can you provide me more details on how you proceed? I have tested on Arch and it seems to work : I have also deleted cache / config folders and tried each time with tracks from library or from File Explorer. Playlists are stored in local sqlite db and successfully restored at startup.

**Comment by Ilya87** [comment]
Demo for this and for https://github.com/MBach/Miam-Player/issues/17 bugs.
https://yadi.sk/i/mViI_b5PfPqZq

**Comment by MBach** [comment]
Great ! (but next time, please put it into English :) 

I still struggle to understand why it's not saving. I suspect the same encoding problem with russian or japanese names (in the filename).

**Comment by Ilya87** [comment]
Note: my settings are to save playlist on exit and to load it on start.
Current behavior: 
1). after creating new playlist (and you have no other in Playlist Manager), playlist can't be saved using Menu ->Quit command. You should use window manager's close button.
2). after you quit once again using window manager's close button and start Miam-Player again, then add some tracks to the first playlist, restart Miam-Player in the same manner, amount of each track in playlist is doubled. And then after each restart of Miam-Player (even without track adding) each track will be doubled once again.
3). after you create second playlist, add tracks there, close Miam-Player using window manager's close button and start again - you'll see second playlist before the first one.
4). now add tracks to the second playlist and restart Miam-Player (you know how) and you'll see second playlist with old and newly added tracks in the beginning, then second playlist with only old tracks (without newly added), and in the end - first playlist.

I also spotted situation when second playlist's tracks were merged into the first one (with disappearing of the second playlist), but it is a bit more difficult to track this behavior.

**Comment by MBach** [comment]
Ok I have seen some errors in logs like:

    int __cdecl SqlDatabase::insertIntoTablePlaylists(const class PlaylistDAO &,const class std::list<class TrackDAO,class std::allocator<class TrackDAO> > &,bool) 
    QSqlError("19", "Impossible to fetch row", "UNIQUE constraint failed: playlists.id")

First point should be fixed anyway: it's possible to automatically save with Exit or with close button.

Merge happens when 1st playlist was saved, then Miam-Player was closed, 1st was restored at startup, then you add a 2nd one. When starting over, 1st and 2nd are merged with wrong hash calculated, which leads to save over and over.

**Comment by Ilya87** [comment]
The second point has changed - after adding new tracks to the first playlist and restart of Miam-Player newly created first playlist with new and old tracks appears, but also remains playlist with old tracks.

**Comment by MBach** [comment]
I changed how random numbers where generated in https://github.com/MBach/Miam-Player/commit/430cbb3087d82e80703ba6134a39672909a8682c. 

Random numbers were different when inserting to database during same launch of Miam-Player (so you could save 2 playlists without problems). However, the sequence was the same every time the program was started, so a collision could happen with a very high probability.

**Comment by Ilya87** [comment]
Now when I create second playlist and restart Miam-Player second playlist becomes clean (no tracks) and even its name also is empty - not standard, but empty.

**Comment by MBach** [comment]
Damn it! I was too fast. I could reproduce the issue. I changed with unsigned int everywhere instead of int and unsigned int, this is not happening anymore for me!
https://github.com/MBach/Miam-Player/commit/33aca2a5951b15ee350ff417063905e29b721c63

**Comment by Ilya87** [comment]
What remains

**Comment by MBach** [comment]
Actually this is the expected behaviour but we can discuss if it's normal or not. When you are closing a playlist that you've modified, a popup shows on screen and ask you if you want to replace it, or discard your changes.

But, when you're exiting the application, I think it's annoying to have this popup at the last minute to chose what you want to do! So I have decided to automatically save modified playlist into a new one. That's why you have 2 playlists: the unmodified before you quit, and the newly saved one.

Maybe this can be improved with the first setting in options:
![playlists_settings](https://cloud.githubusercontent.com/assets/1532657/7754245/6def9068-ffeb-11e4-979e-eb7ac6393c7d.png)
* If "Let me decide" is checked, then you cannot exit application before making a choice
* If "Save" is checked, then every modified playlists are overwritten
* If "Discard" is checked, then every modification is lost

What do you think?

**Comment by MBach** [comment]
Another point, if such popup is showing, and if you have 3 modified playlists, you will have 3 times the popup. One solution could be to create a special popup with "Save all" or "Discard all" or "Cancel". But it requires some new code :)

**Comment by Ilya87** [comment]
I don't really understand the aim of the current behavior: I am the person who added tracks to playlist. I could add them to new playlist, but I have chosen to add to already existing playlist. This was my choice. 
If I would like to create a copy of already existing playlist I would make a right mouse button click on playlist tab an selected point - «Create a copy of playlist». And this is the right behavior IMO.

**Comment by MBach** [comment]
Good point, making a copy of playlist without telling user explicitly can be very disturbing. 

So I will follow the comment I made earlier. Depending how "Closing policy" is configured, I will implement 3 ways to exit application with modified playlist : with popup, overwritting and discarding changes.

**Comment by MBach** [comment]
I hope this one is the one! 
https://github.com/MBach/Miam-Player/commit/61ab79ac93bc565beb343e415365d2ba376e131f
But it's quite long to test every use case ...

**Comment by Ilya87** [comment]
1 - Now when I exit and "Save" field is checked Miam-Player asks me what to do.
2 - Second playlist moves after restart before the first one in playlist tabbar.

It's my first 1 min. overview.

**Comment by MBach** [comment]
Try this one: https://github.com/MBach/Miam-Player/commit/003589a94259ca65b60c715418633b7ed469c049
I have redesigned how playlists are managed. It wasn't clear to mix a "Closing Policy". Lots of stranges cases needed to be implemented, and the User Experience can be very confusing.

So far, here is what I have made:
![playlists_options](https://cloud.githubusercontent.com/assets/1532657/8120452/86b43472-109f-11e5-9940-945dfc0904c6.png)
In a group box, options are separated to indicate they cannot interact each others.

Also, in the playlist manager, now you can see which playlists are saved and modified:
![modified_playlist](https://cloud.githubusercontent.com/assets/1532657/8120508/226e452e-10a0-11e5-8340-695d47fdcace.png)
A fourth button to save modified should be enabled only for modified ones.

**Comment by Ilya87** [comment]
When I rename playlists and exit then restart - playlist name became the default.

**Comment by Ilya87** [comment]
Also closed second playlist after restart appears again. And I should manually delete this playlist from the list.

**Comment by MBach** [comment]
1. Default name should be fixed.
2. I didn't understand the second point? Are you sure that table Playlists and PlaylistTracks (you can use http://sqlitebrowser.org/ or any tool to visualize database file mp.db)?

**Comment by Ilya87** [comment]
About second point - I mean you create new playlist (in my case it is the second), then restart (close and start) Miam Player, then close plylist using cross icon on it, then restart Miam-Player again - second playlist will be loaded, because after closing it remains in Playlist manager.

**Comment by Ilya87** [comment]
From the other hand deleting playlist from Playlist manager closes it in playlist bar.

**Comment by MBach** [comment]
Well if you have chosen to load automatically playlists at startup, no matter if you close them manually or not: next time Miam-Player will open, it load them again!
I think it's time to close this issue :)?

**Comment by Ilya87** [comment]
Logically yes, but this behavior is a little bit unusual. Would you please be so kind to add delete option to playlist contex menu to save time a little - no need to go to playlist manager, search for necessary playlist and only after that delete it.

**Comment by MBach** [comment]
Ah OK I see what you would like to have. 
Here is what I understand from you: only opened playlists should be restored at startup, not all the playlists possibly stored in database (which can be huge)! It's like when you want to restore last session in Firefox or in Qt Creator in fact.

It might be a more common way to restore playlists.

Finally, the Delete item in the context menu is also a good idea: it can be a waste of time to force the user to open Playlist Manager.

**Comment by MBach** [comment]
https://github.com/MBach/Miam-Player/commit/817ccdde3b347d6ef060ba225044db2be184e0b9
According to your comments, I have changed how playlists should be loaded at startup.
Moreover, tab order is implemented: previous selected tab and playlist order should be preserved!

**Comment by MBach** [comment]
And delete option is included in the context menu, with explicit confirm from user (that's why it's labelled "Delete playlist..." with "...")

**Comment by Ilya87** [comment]
WOW! It works - all closed and not deleted playlist are stayed in Playlist Manager, and not restored at startup. Thank you.
